src/platforms/virtual/platform.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+#include "display.h"
+#include "mir/graphics/platform.h"
+#include "options_parsing_helpers.h"
+
+namespace mg = mir::graphics;
+namespace mgv = mir::graphics::virt;
+namespace geom = mir::geometry;
+using namespace std::literals;
+
+
+class mgv::Platform::VirtualDisplayInterfaceProvider : public mg::DisplayInterfaceProvider
+{
+public:
+    explicit VirtualDisplayInterfaceProvider()
+    {
+    }
+
+protected:
+    auto maybe_create_interface(mg::DisplayProvider::Tag const& type_tag)
+        -> std::shared_ptr<mg::DisplayProvider>
+    {
+        class StubGenericEGLDisplayProvider : public GenericEGLDisplayProvider
+        {
+        public:
+            auto get_egl_display() -> EGLDisplay override
+            {
+                auto egl_display = eglGetDisplay(EGL_DEFAULT_DISPLAY);
+                EGLint major, minor;
+                if (eglInitialize(egl_display, &major, &minor) == EGL_FALSE)
+                    BOOST_THROW_EXCEPTION(std::runtime_error("Failed to initialize EGL display"));
+                return egl_display;
+            }
+
+            class StubEGLFramebuffer : public EGLFramebuffer
+            {
+            public:
+                StubEGLFramebuffer() = default;
+                StubEGLFramebuffer(StubEGLFramebuffer const&) {}
+                void make_current() override {}
+                void release_current() override {};
+                auto clone_handle() -> std::unique_ptr<EGLFramebuffer>  override
+                {
+                    return std::make_unique<StubEGLFramebuffer>(*this);
+                }
+                auto size() const -> geom::Size override
+                {
+                    return geom::Size{};
+                }
+            };
+
+            auto alloc_framebuffer(GLConfig const&, EGLContext) -> std::unique_ptr<EGLFramebuffer> override
+            {
+                return std::make_unique<StubEGLFramebuffer>();
+            }
+
+        private:
+        };
+
+        if (dynamic_cast<mg::GenericEGLDisplayProvider::Tag const*>(&type_tag))
+        {
+            return std::make_shared<StubGenericEGLDisplayProvider>();
+        }
```

Comment:
I don't think we _can_ return a viable GenericEGLDisplayProvider, so we should just return an empty `std::shared_ptr<mg::DisplayProvider>`.

That allows the rendering provider to allocate from another platform

Reply:
What type of `DisplayProvider` do you think should be returned here? The reason for supplying a `GenericEGLDisplayProvider` is that it makes the `BasicScreenshooter` work, which is required for VNC.

Reply:
I hadn't realised there was code that needed this.

While trying to understand the error I encountered (above, trying to use the virtual platform on its own) I was looking at `egl_display_from_platforms()` in the generic egl renderer. That looks through the displays to find a provider that can offer an `EGLDisplay`, but this code fails because `eglInitialize()`  doesn't work. It just seems logical not to claim we support generic EGL when it doesn't work which would allow `egl_display_from_platforms()` to use another platform.

Reply:
Okay I think I understand you better now. So if the "virtual" platform's EGLDisplay fails _and_ we are in a hybrid situation, we _might_ be able to back up to the interface provided by another platform.

I think we still need to try to create an EGLDisplay on the purely virtual platform so that we can support that case (since there would be nothing to fall back to).

I am unsure if there are any consequences of this though, but I can test

Reply:
If `eglInitialize` fails altogether though, we are in a bit of a situation. That would basically mean that we don't have a rendering platform (or we could have the `NullRenderingPlatform`). And nothing would function correctly in this scenario AFAICT.

Reply:
There's a whole lot of "try something and hope it works" we likely need to deal with (outside of this PR). So, if this is making something work, I don't insist on getting rid. I think I just need to better understand what works, when and why.

Reply:
> If `eglInitialize` fails altogether though, we are in a bit of a situation.

If `eglInitialize` fails we should *not* return a `GenericEGLDisplayProvider`, because anything that tries to use it will fail.

What we *could* do is provide a `CPUAddressableDisplayProvider`; that would let other `RenderingPlatform`s (gbm-kms or eglstream-kms) load.